_source/_posts/2021-11-08-k8s-api-server-oidc.md

```diff
@@ -0,0 +1,565 @@
+
+## Set up an Okta OIDC application and authorization server
+
+You can achieve OIDC login for the cluster by creating a simple OIDC application with Okta either using the Okta CLI or the Admin Console. But with an OIDC application alone, you would have to use the client secret to authenticate from kubectl or any other client library. Which does not scale and is not better than default k8s authentication mechanisms as you won't have granular controls over users and roles. For a more helpful setup, we would need an OIDC application and an authorization server with customized claims and policies for Kubernetes. This way, we can make use of Okta to manage users and permissions as well.
+
+There are multiple ways to set up an OIDC application and authorization server in Okta. If you prefer to do this via a GUI, then follow the **Configure Your Okta Org section** from [this article](/blog/2021/10/08/secure-access-to-aws-eks#configure-your-okta-org) to do it via the Okta Admin Console.
+
+In this tutorial, we will use Terraform to configure the Okta part so that you can reuse the code for any automation required. Let's dive into each step needed.
+
+### Set up Terraform
+
+If you don't have Terraform installed, follow [these instructions](https://learn.hashicorp.com/tutorials/terraform/install-cli) and set it up.
+
+You can find the complete Terraform source code for this article in this [GitHub repo](https://github.com/oktadev/okta-k8s-oidc-terraform-example)
+
+First, we need to configure the [Okta Terraform provider](https://registry.terraform.io/providers/okta/okta/latest/docs). Create a new Terraform file, let's say `okta.main.tf`, and add the following:
+
+```hcl
+variable "base_url" {
+  description = "The Okta base URL. Example: okta.com, oktapreview.com, etc. This is the domain part of your Okta org URL"
+}
+variable "org_name" {
+  description = "The Okta org name. This is the part before the domain in your Okta org URL"
+}
+variable "api_token" {
+  type        = string
+  description = "The Okta API token, this will be read from environment variable (TF_VAR_api_token) for security"
+  sensitive   = true
+}
+
+# Enable and configure the Okta provider
+terraform {
+  required_providers {
+    okta = {
+      source  = "okta/okta"
+      version = "~> 3.15"
+    }
+  }
+}
+
+provider "okta" {
+  org_name  = var.org_name
+  base_url  = var.base_url
+  api_token = var.api_token
+}
+```
+
+You need to provide the input variables `org_name`, `base_url`, and `api_token`. For example, if the address of your Okta instance is `dev-1234.okta.com`, then your `org_name` would be `dev-1234`, and `base_url` would be everything that comes after the org name (e.g., okta.com). Update these values in a file named `okta.config.auto.tfvars`. The `.auto` in the name is important; otherwise, Terraform will not pick it up automatically.
```

Comment:
```suggestion
You need to provide the input variables `org_name`, `base_url`, and `api_token`. Update these values in a file named `okta.config.auto.tfvars`. The `.auto` in the name is important; otherwise, Terraform will not pick it up automatically. For example, if the address of your Okta instance is `dev-1234.okta.com`, then your `org_name` would be `dev-1234`, and `base_url` would be everything that comes after the org name (e.g., okta.com). 
```